orca-mahe/src/main/groovy/com/netflix/spinnaker/orca/mahe/cleanup/FastPropertyCleanupListener.groovy

```diff
@@ -55,34 +58,82 @@ class FastPropertyCleanupListener implements ExecutionListener {
       rollbacks.each { stage ->
         switch (stage.context.propertyAction) {
           case PropertyAction.CREATE.toString():
-            stage.context.propertyIdList.each { prop ->
-              log.info("Rolling back the creation of: ${prop.propertyId} on execution ${execution.id} by deleting")
-              Response response = mahe.deleteProperty(prop.propertyId, "spinnaker rollback", extractEnvironment(prop.propertyId))
-              resolveRollbackResponse(response, stage.context.propertyAction.toString(), prop)
+            stage.context.persistedProperties.each { Map prop ->
+              String propertyId = prop.propertyId
+              if (shouldRollback(prop)) {
+                log.info("Rolling back the creation of: ${propertyId} on execution ${execution.id} by deleting")
+                Response response = mahe.deleteProperty(propertyId, "spinnaker rollback", extractEnvironment(propertyId))
+                resolveRollbackResponse(response, stage.context.propertyAction.toString(), prop)
+              } else {
+                log.info("Property has been updated since this execution; not rolling back")
+              }
             }
             break
           case PropertyAction.UPDATE.toString():
-            stage.context.originalProperties.each { prop ->
-              log.info("Rolling back the ${stage.context.propertyAction} of: ${prop.property.propertyId} on execution ${execution.id} by upserting")
-              Response response = mahe.upsertProperty(prop)
-              resolveRollbackResponse(response, stage.context.propertyAction.toString(), prop.property)
+            stage.context.originalProperties.each { Map originalProp ->
+              Map property = originalProp.property
+              Map updatedProperty = (Map) stage.context.persistedProperties.find { it.propertyId == property.propertyId }
+              String propertyId = property.propertyId
+              if (shouldRollback(updatedProperty)) {
+                log.info("Rolling back the update of: ${propertyId} on execution ${execution.id} by upserting")
+                Response response = mahe.upsertProperty(originalProp)
+                resolveRollbackResponse(response, stage.context.propertyAction.toString(), property)
+              } else {
+                log.info("Property has been updated since this execution; not rolling back")
```

Comment:
Should we actually make the pipeline fail?

Reply:
Good question. Can we actually make the pipeline fail in a listener like this? If we can, how can we surface the reason for it failing to the user?